utils/dbutil/autocompact/store.go

```diff
@@ -0,0 +1,124 @@
+package autocompact
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/Fantom-foundation/lachesis-base/kvdb"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type Store struct {
+	kvdb.Store
+	minKey  []byte
+	maxKey  []byte
+	written uint64
+	limit   uint64
+	compMu  sync.Mutex
+}
+
+type Batch struct {
+	kvdb.Batch
+	written uint64
+	minKey  []byte
+	maxKey  []byte
+	onWrite func(key []byte, size uint64, force bool)
+}
+
+func Wrap(s kvdb.Store, limit uint64) *Store {
+	return &Store{
+		Store: s,
+		limit: limit,
+	}
+}
+
+func (s *Store) onWrite(key []byte, size uint64, force bool) {
+	s.compMu.Lock()
+	defer s.compMu.Unlock()
+	s.written += size
+	if s.minKey == nil {
+		s.minKey = common.CopyBytes(key)
+		s.maxKey = common.CopyBytes(key)
+	}
+	if key != nil && bytes.Compare(key, s.minKey) < 0 {
+		s.minKey = common.CopyBytes(key)
+	}
+	if key != nil && bytes.Compare(key, s.maxKey) > 0 {
+		s.maxKey = common.CopyBytes(key)
+	}
+	if force || s.written > s.limit {
+		_ = s.Store.Compact(s.minKey, s.maxKey)
+		s.written = 0
+		s.minKey = nil
+		s.maxKey = nil
+	}
+}
+
+func (s *Store) Put(key []byte, value []byte) error {
+	defer s.onWrite(key, uint64(len(key)+len(value)+64), false)
+	return s.Store.Put(key, value)
+}
+
+func (s *Store) Delete(key []byte) error {
+	defer s.onWrite(key, uint64(len(key)+64), false)
+	return s.Store.Delete(key)
+}
+
+func (s *Store) Close() error {
+	s.onWrite(nil, 0, true)
+	return s.Store.Close()
+}
+
+func (s *Store) NewBatch() kvdb.Batch {
+	batch := s.Store.NewBatch()
+	if batch == nil {
+		return nil
+	}
+	return &Batch{
+		Batch:   batch,
+		onWrite: s.onWrite,
+	}
+}
+
+func (s *Batch) Put(key []byte, value []byte) error {
+	s.written += uint64(len(key) + len(value) + 64)
```

Comment:
Can you explain, please, why you add up 64 here?

Reply:
Storage overheads, related to adding/deleting a record, wouldn't be only proportional to length of key and value. E.g. if you one adds 10 records with length of `2`, it will be more expensive than 1 record with length `20`
Now, `64` wasn't really calculated but is rather a guesstimation